Added next milestone display

TheHollidayInn opened the pull request:
Resolves #5057 .

I have added a bar that lists the progress to and the next milestone. Should I add the text to the translation files and do we want a progress bar to appear? And, do we want to link to the milestone on a wiki page for more explanation?

Also, I have added tests.

Here is a screen shot:

![screen shot 2015-04-22 at 10 43 11 am](https://cloud.githubusercontent.com/assets/1253400/7278734/efd2e344-e8dc-11e4-8dfa-48f0f84be963.png)

TheHollidayInn committed:
Added next milestone display

crookedneighbor commented on test/spec/headerCtrlSpec.js:
```diff
@@ -0,0 +1,72 @@
+
+    scope.user.stats.lvl = 1;
+    scope.getNextMilestone();
+    expect(scope.nextMilestone).eql({"level": '4',  text: "Pets + Mounts Unlocked"});
```
Is there a reason level is in quotes while text is not?

TheHollidayInn replied:
Nope, I believe the quotes can be removed. I'll make the change and rerun the tests.

crookedneighbor replied:
Cool. Either way is fine, it should just be consistent. :smile:

crookedneighbor replied:
Also, big :+1: for including tests.

crookedneighbor commented:
Can you submit a screenshot of how it looks on a smaller screen size? I'm particularly concerned about wrapping issues with the longer strings of text. 

@lemoness Is this where you wanted the milestones meter?

crookedneighbor commented:
In answer to your question about translations, yes, we will want these strings to be translatable, but I'd wait till the design and placement is approved before doing the work of creating translatable strings.

AnitaYCheng commented:
My concern about this solution is that the role of the progress bar doesn't match the rest of the bars. The other 3 bars mark progress _within_ the level, to so say, so the units are a lot smaller and they visually change more readily. The milestone bar's units are "level" which will go up a lot more slowly than the others. Since they have different roles, they should be grouped separately.

Although I'm not quite sure where it should go if not in the header with the other bars. Just wanted to add my thoughts about grouping.

lemoness commented:
Personally, I wasn't envisioning it as a bar, for the reasons @sugarfiend
just mentioned - just as a display. As an example of one of the things that
confused me: since the previous milestone before the Free Orb of Rebirth is
at level 60, is it counting from level 60 to level 100? Or does it start at
60?

I'd prefer it just say something like NEXT MILESTONE at Level 100: Free Orb
of Rebirth!

A visual reminder, not a counter - since users can already see their levels
go up.

On Wed, Apr 22, 2015 at 12:49 PM, anita notifications@github.com wrote:

> My concern about this solution is that the role of the progress bar
> doesn't match the rest of the bars. The other 3 bars mark progress
> _within_ the level, to so say, so the units are a lot smaller and they
> visually change more readily. The milestone bar's units are "level" which
> will go up a lot more slowly than the others. Since they have different
> roles, they should be grouped separately.
> 
> Although I'm not quite sure where it should go if not in the header with
> the other bars. Just wanted to add my thoughts about grouping.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-95315342.

TheHollidayInn committed:
Removed bar, add locals, and fixed tests

TheHollidayInn commented:
Is this more of what you are looking for? Also, I fixed the tests and added locals.

![screen shot 2015-05-08 at 7 10 57 pm](https://cloud.githubusercontent.com/assets/1253400/7547967/6e6b5fe0-f5be-11e4-8d0f-c1456b8eaf49.png)
![screen shot 2015-05-08 at 7 11 11 pm](https://cloud.githubusercontent.com/assets/1253400/7547968/6e6f89a8-f5be-11e4-950b-9e9ed5083067.png)

crookedneighbor commented:
I'm confused about why it says "Next milestone at level 69: Free Orb of Rebirth". The Orb becomes free at level 100.

TheHollidayInn committed:
Fixed level display and formatting issue

TheHollidayInn commented:
Eh, my bad. I was using the user level. 

![screen shot 2015-05-08 at 9 06 15 pm](https://cloud.githubusercontent.com/assets/1253400/7548208/474d1ac2-f5c6-11e4-919b-289f7c0e520e.png)

![screen shot 2015-05-08 at 9 06 20 pm](https://cloud.githubusercontent.com/assets/1253400/7548207/35a18d44-f5c6-11e4-9ea6-bfaabb31683d.png)

lemoness commented:
Yes, that's much more what I was looking for! @anitaycheng, any additional
thoughts?

On Fri, May 8, 2015 at 7:08 PM, Keith Holliday notifications@github.com
wrote:

> Eh, my bad. I was using the user level.
> 
> [image: screen shot 2015-05-08 at 9 06 15 pm]
> https://cloud.githubusercontent.com/assets/1253400/7548208/474d1ac2-f5c6-11e4-919b-289f7c0e520e.png
> 
> [image: screen shot 2015-05-08 at 9 06 20 pm]
> https://cloud.githubusercontent.com/assets/1253400/7548207/35a18d44-f5c6-11e4-9ea6-bfaabb31683d.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-100411540.

deilann commented:
I don't like it. It doesn't match the style of the header at all, it's
off-alignment with the other information, and it provides spoilers that
users might not want to see. I don't think it should _tell_ you what the
milestone is, like I think @Alys stated. It should also definitely be
incredibly easy to hide and should recognize that once that milestone is
reached, if a user dies to a level below it, the milestone stays the same.
Should be looking at flags, not user level.

On Wed, May 13, 2015 at 3:05 PM, lemoness notifications@github.com wrote:

> Yes, that's much more what I was looking for! @anitaycheng, any additional
> thoughts?
> 
> On Fri, May 8, 2015 at 7:08 PM, Keith Holliday notifications@github.com
> wrote:
> 
> > Eh, my bad. I was using the user level.
> > 
> > [image: screen shot 2015-05-08 at 9 06 15 pm]
> > <
> > https://cloud.githubusercontent.com/assets/1253400/7548208/474d1ac2-f5c6-11e4-919b-289f7c0e520e.png
> > 
> > [image: screen shot 2015-05-08 at 9 06 20 pm]
> > <
> > https://cloud.githubusercontent.com/assets/1253400/7548207/35a18d44-f5c6-11e4-9ea6-bfaabb31683d.png
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-100411540.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-101830064.

TheHollidayInn commented:
I can add the hide, but I thought we were trying to limit the user settings. I just added another user setting with the add party button. Either way, just let me know.

deilann commented:
From what I can read of the code, this still doesn't address the issue that this shouldn't be based off of user level. If a user hits level 30, then dies twice, then levels up again and is level 29, their next milestone isn't at level 30. It's at level 40.

And yes, we want to reduce user settings, but this is really not pleasing to the eye, makes the header bigger, contains spoilers for new users, and information older users don't need. I'd be incredibly annoyed if I couldn't hide it.

crookedneighbor commented:
I always imagined this feature on the stats page. Not in the header.

deilann commented:
If we put it there it could also be "click to see next milestone"

On Wed, May 13, 2015 at 9:13 PM, Blade Barringer notifications@github.com
wrote:

> I always imagined this feature on the stats page. Not in the header.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-101904147.

Alys commented:
There could also be "click to see all milestones"

TheHollidayInn commented:
Ah, @deilann , that's what you meant by flags. I can do that. 

I personally like this in the header, because I think people should see their progress or have suggestions on what else the game has to offer.

@Alys I like the idea.

Either way, I don't have strong opinions, so just let me know what you want me to do :)

SabreCat commented:
Stats page seems like a good place to put it.

deilann commented:
If it's going to be put in the stats page (which I think is honestly the
best place for it), let's mention it in the tutorial. I think only the
level 10 tutorial takes you to the stats page, though, and I think this
needs to be made clear beforehand.

Maybe instead of actually taking you to the stats page, the original
tutorial can mention that it's there, something like this at the end:

"Wondering what comes next? If you want to spoil the surprise, check out
the stats page. You can view your next milestone there... or all the
milestones Habit has to offer!"

lemoness commented:
I do think that it's important to have a user setting that will disable the
milestone, or maybe a "click to hide, click to open" - as @deilann says,
many preexisting users won't want or need it. As long as it isn't too easy
to accidentally hide, or is easy to show again if you hide it accidentally,
since based on the user onboarding tests we've been running, new users
click wildly on everything.

I don't mind also having a version in the stats page, but having it ONLY in
the stats page would defeat the purpose - we want the fact that there are
upcoming things to be very obvious! That being said, I think having a
version on the stats page would be valuable, because we could include much
more detail. I'm hesitant to rely on the tutorial, because we are already
having a very tough time optimizing it so that users actually read and
understand it.

Re: showing what the milestone is, versus simply showing that there is a
milestone... I'm concerned about removing the name of the milestone,
because right now we don't have very good indications of when that
milestone has been achieved, especially for the quests. We plan to make
that more obvious in the future, but until then, I'd prefer to have the
name of the upcoming milestone be included. In the user onboarding tests,
we see excitement when people "discover" features by digging deeper - "Oh
wow, that tab says, 'Pets'! There are going to be pets? Cool, I want one,
how do I get it..." Unfortunately, many people never dig, and never get to
feel that excitement.

As for changing the UI to make it fit better, I definitely think it could
be improved so that it's a better match to the stat bar - maybe colored
green, to go with the red, yellow, and blue? Or simply adjusted?

On Thu, May 14, 2015 at 4:28 PM, deilann notifications@github.com wrote:

> If it's going to be put in the stats page (which I think is honestly the
> best place for it), let's mention it in the tutorial. I think only the
> level 10 tutorial takes you to the stats page, though, and I think this
> needs to be made clear beforehand.
> 
> Maybe instead of actually taking you to the stats page, the original
> tutorial can mention that it's there, something like this at the end:
> 
> "Wondering what comes next? If you want to spoil the surprise, check out
> the stats page. You can view your next milestone there... or all the
> milestones Habit has to offer!"
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-102198758.

TheHollidayInn commented:
Okay, so in summary here are the changes proposed:
-Add a hide milestone message
-Add a view all upcoming milestones
-(maybe) move milestones to stats page
-Make milestone logic based on user flags rather than level

As for changing the UI, I had a green progress bar at the top that seemed to match closer. Maybe we can start from that direction.

lemoness commented:
@AnitaYCheng Do you want to do any mockups that might make the milestones feature look better while in the header?

veeeeeee commented:
I think more visual stuff might complicate this part of the screen. Suggest plain text (no highlighted background), like how gems and gold in the header bar are plain text numbers with no box to enclose them. 

Rather than make it easy to hide, have an option to allow users to input their own goal that they're working towards or other text. The users who don't want to deal with it can leave it blank; others can put in a main goal (ex. finish graduate thesis) or even an inspirational quote per their own preferences. The milestones are simply the default "goals" to let users have something to aim at.

Alys commented:
_"Rather than make it easy to hide, have an option to allow users to input their own goal that they're working towards or other text."_  Ohhhhh! That would be so popular! 

Have a little pencil edit icon next to the text, which opens an edit modal. The modal could have a button to restore the default milestone (for whatever the next milestone is for that user). Perhaps another button to show all milestones. 

Perhaps also a button labelled something like "show nothing" which puts a blank space in the text field and closes the modal (it's both a shortcut for doing that manually, and a way of telling the user that they can hide the message altogether if they want).

crookedneighbor commented:
I'm 1000% on board with this idea now.

deilann commented:
Sooooo am I.

On Mon, May 18, 2015 at 8:28 PM, Blade Barringer notifications@github.com
wrote:

> I'm 1000% on board with this idea now.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-103318516.

TheHollidayInn commented:
I'm pretty confused on this now. If a user enters a goal, how do we track if the reach the goal?

hairlessbear commented:
@TheHollidayInn We don't. User-entered milestones would be there as a reminder of what they're working towards. Users will presumably change or remove their milestone once they reach it.

TheHollidayInn commented:
So, I wouldn't show the next milestone at level xx if they have a custom milestone?

hairlessbear commented:
I assume so, but you'll need to wait for one of the admins for confirmation.

lemoness commented:
This does seem like two slightly different features, but if we can find a
good way to combine them, I would be amenable. Is it possible to have the
in-game milestones implemented first, and then implement the personal
milestones afterwards?

On Tue, May 19, 2015 at 7:10 PM, hairlessbear notifications@github.com
wrote:

> I assume so, but you'll need to wait for one of the admins for
> confirmation.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-103724294.

deilann commented:
If we implement both at the same time, we have no reason to have a hide
feature and won't annoy older users with the feature, while having
something useful for them.

On Wed, May 20, 2015 at 1:28 PM, lemoness notifications@github.com wrote:

> This does seem like two slightly different features, but if we can find a
> good way to combine them, I would be amenable. Is it possible to have the
> in-game milestones implemented first, and then implement the personal
> milestones afterwards?
> 
> On Tue, May 19, 2015 at 7:10 PM, hairlessbear notifications@github.com
> wrote:
> 
> > I assume so, but you'll need to wait for one of the admins for
> > confirmation.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-103724294.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-104026033.

lemoness commented:
I'm no designer, so anyone definitely feel free to edit these as you see fit, but here are some rough mockups of the new idea, combining milestones and goals:

v1
![milestones mockups1a](https://cloud.githubusercontent.com/assets/3494081/7740338/c1e5b890-ff24-11e4-9e5f-ec7bbd343fa9.png)
![milestones mockups1b](https://cloud.githubusercontent.com/assets/3494081/7740345/d3626898-ff24-11e4-87b7-c92aa76e691f.png)

v2
![milestones mockups2a](https://cloud.githubusercontent.com/assets/3494081/7740347/d8f485b6-ff24-11e4-923e-97cdd6a88d9d.png)
![milestones mockups2b](https://cloud.githubusercontent.com/assets/3494081/7740348/dc4c171a-ff24-11e4-8179-20298ca2be10.png)

deilann commented:
I think it'd look better without a bar.

lemoness commented:
Like this? Or no divider?
![milestones mockups3a](https://cloud.githubusercontent.com/assets/3494081/7765430/a62f36be-0012-11e5-9ef8-2e59cb15d83c.png)

lemoness commented:
Would also love some feedback from @AnitaYCheng @veeeeeee about these options!

SabreCat commented:
No divider, I'd say. Looks nice!

On Fri, May 22, 2015, 1:40 AM lemoness notifications@github.com wrote:

> Would also love some feedback from @AnitaYCheng
> https://github.com/AnitaYCheng @veeeeeee https://github.com/veeeeeee
> about these options!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-104538022.

lemoness commented:
I personally dislike it with just text, but I am willing to be overruled!

veeeeeee commented:
Definitely no divider, and try it plaintext (not bolded) as well.

lemoness commented:
Okay, let's go with this then (using the normal site font, since this is just the one I had in my editor). A designer can always improve it later.

![milestones mockups](https://cloud.githubusercontent.com/assets/3494081/7822276/4f37e316-03a9-11e5-9c74-ca7e36b889a1.png)

TheHollidayInn commented:
Will we still be adding the hide message and the view all message?

lemoness commented:
View-all message could be useful, although I would like to see screenshots.

On Tue, May 26, 2015 at 8:27 PM, Keith Holliday notifications@github.com
wrote:

> Will we still be adding the hide message and the view all message?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-105743037.

TheHollidayInn committed:
Added hide next milestone message setting

TheHollidayInn committed:
Added View all milesones modal

TheHollidayInn committed:
Adjusted display of next milestone message

TheHollidayInn commented:
![screen shot 2015-05-30 at 11 17 45 am](https://cloud.githubusercontent.com/assets/1253400/7898195/d8893c84-06bd-11e5-8839-d9de3b9ce9ef.png)

![screen shot 2015-05-30 at 11 17 39 am](https://cloud.githubusercontent.com/assets/1253400/7898194/d8862d78-06bd-11e5-9cb6-b60290f6c558.png)

deilann commented:
I really don't want this to make the header bigger. It looks awful to have
the header bigger than the avatar, and on many resolutions, the header
already takes up a substantial amount of room. I don't think we should do
anything that makes it take more.

On Sat, May 30, 2015 at 9:20 AM, Keith Holliday notifications@github.com
wrote:

> [image: screen shot 2015-05-30 at 11 17 45 am]
> https://cloud.githubusercontent.com/assets/1253400/7898195/d8893c84-06bd-11e5-8839-d9de3b9ce9ef.png
> 
> [image: screen shot 2015-05-30 at 11 17 39 am]
> https://cloud.githubusercontent.com/assets/1253400/7898194/d8862d78-06bd-11e5-9cb6-b60290f6c558.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5062#issuecomment-107059055.

crookedneighbor commented:
I agree with @deilann. We probably need to enlist a designer to figure out how to present this feature.

crookedneighbor commented:
We've decided to table this feature until we have a new app design in place. Will close this now, but if we use any of your commits in the future for this feature, you will be credited.
